packages/ckb-sdk-core/src/transactionBuilder.ts

```diff
@@ -0,0 +1,64 @@
+import { rawTransactionToHash } from '@nervosnetwork/ckb-sdk-utils'
+import { ArgumentRequired } from '@nervosnetwork/ckb-sdk-utils/lib/exceptions'
+import signWitness from './signWitness'
+import Address from './address'
+
+class TransactionBuilder {
+  public hash: CKBComponents.Hash256 = ''
+
+  public rawTransaction: CKBComponents.RawTransaction = {
+    version: '0x0',
+    cellDeps: [],
+    headerDeps: [],
+    inputs: [],
+    outputs: [],
+    witnesses: [],
+    outputsData: [],
+  }
+
+  constructor({ inputs = [], outputs = [], ...rest }: TransactionBuilderInitParams = { inputs: [], outputs: [] }) {
+    this.rawTransaction = {
+      ...this.rawTransaction,
+      ...rest,
+      inputs,
+      outputs,
+    }
+  }
+
+  public addInput = (input: CKBComponents.CellInput) => this.rawTransaction.inputs.push(input)
+
+  public addOutput = (output: CKBComponents.CellOutput) => this.rawTransaction.outputs.push(output)
+
+  public removeInput = (index: number) => this.rawTransaction.inputs.splice(index, 1)
+
+  public removeOutput = (index: number) => this.rawTransaction.outputs.splice(index, 1)
+
+  public updateOptions = (options: Omit<CKBComponents.RawTransaction, 'inputs' | 'outputs'>) => {
+    this.rawTransaction = {
+      ...this.rawTransaction,
+      ...options,
+    }
+  }
+
+  public updateTransactionHash = () => {
```

Comment:
When and where `updateTransactionHash ` function will be called? I don't catch it and I think it could be private.

Reply:
It will be called automatically on calling `signInput` if the `builder.hash` is [empty](https://github.com/nervosnetwork/ckb-sdk-js/pull/362/files/d5d69c014c86d645e3dffae686fb12fe83956f1e#diff-79abb8d630ddf83d5a63cd7efa02fb86R56). But it should be able to update the has manually if the raw transaction is updated.

Reply:
OK. I think sometimes we support more interfaces maybe not a good thing.